client/my-sites/stats/activity-log-tasklist/index.jsx

```diff
@@ -0,0 +1,414 @@
+		let plugins;
+
+		if ( 'string' === typeof pluginSlug ) {
+			plugins = [ pluginSlug ];
+			trackDismissPlugin( pluginSlug );
+		} else {
+			plugins = Object.keys( pluginsWithUpdate );
+			trackDismissPluginAll();
+		}
+
+		this.setState( {
+			dismissedPlugins: union( this.state.dismissedPlugins, plugins ),
+		} );
+	};
+
+	/**
+	 * Goes to general plugin management screen.
+	 *
+	 * @returns {object} Action to redirect to plugins management.
+	 */
+	goManagePlugins = () => this.props.goManagePlugins( this.props.siteSlug );
+
+	/**
+	 * Goes to single plugin management screen.
+	 *
+	 * @param {string} pluginSlug Plugin slug, like "hello-dolly".
+	 *
+	 * @returns {object} Action to redirect to plugin management.
+	 */
+	goToPlugin = pluginSlug => this.props.goToPlugin( pluginSlug, this.props.siteSlug );
+
+	/**
+	 * Checks if the plugin update queue has more items and none is currently updating.
+	 * If so, updates the next plugin.
+	 */
+	continueQueue = () => {
+		if (
+			0 < this.state.queuedPlugins.length &&
+			! isPluginUpdating( Object.values( this.props.pluginsWithUpdate ) )
+		) {
+			this.updatePlugin( this.state.queuedPlugins[ 0 ] );
+		}
+	};
+
+	/**
+	 * Add a plugin to the update queue.
+	 *
+	 * @param {object} plugin Plugin to enqueue.
+	 *
+	 * @returns {undefined}
+	 */
+	enqueuePlugin = plugin =>
+		this.setState(
+			{
+				queuedPlugins: union( this.state.queuedPlugins, [ plugin ] ),
+			},
+			this.continueQueue
+		);
+
+	/**
+	 * Remove a plugin from the update queue.
+	 *
+	 * @returns {undefined}
+	 */
+	dequeuePlugin = () =>
+		this.setState(
+			{
+				queuedPlugins: drop( this.state.queuedPlugins ),
+			},
+			this.continueQueue
+		);
+
+	/**
+	 * Add all plugins with pending updates to the queue and process it.
+	 *
+	 * @returns {undefined}
+	 */
+	updateAll = () =>
+		this.setState(
+			{
+				queuedPlugins: union(
+					this.state.queuedPlugins,
+					Object.values( this.props.pluginsWithUpdate )
+				),
+			},
+			this.continueQueue
+		);
+
+	/**
+	 * Starts the update process for a specified plugin. Displays an informational notice.
+	 *
+	 * @param {object} plugin Plugin information that includes
+	 * {
+	 * 		{string} id   Plugin id, like "hello-dolly/hello".
+	 * 		{string} slug Plugin slug, like "hello-dolly".
+	 * 		{string} name Plugin name, like "Hello Dolly".
+	 * }
+	 * @param {string} location Send 'from_task' when this is called a row in the task list,
+	 *                          'from_notice' when it's called from error notice.
+	 */
+	updatePlugin = ( plugin, location = 'from_task' ) => {
+		const { showInfoNotice, siteName, updateSinglePlugin } = this.props;
+
+		updateSinglePlugin( plugin, location );
+
+		showInfoNotice(
+			PluginNotices.inProgressMessage( 'UPDATE_PLUGIN', '1 site 1 plugin', {
+				plugin: plugin.name,
+				site: siteName,
+			} ),
+			{
+				id: plugin.slug,
+				showDismiss: false,
+			}
+		);
+	};
+
+	componentDidMount() {
+		page.exit( '*', ( context, next ) => {
```

Comment:
This was taken from 
https://github.com/Automattic/wp-calypso/blob/4b11becda3fc7dec39f5a87832bb2175ebd1e3c2/client/my-sites/plugins/jetpack-plugins-setup/index.jsx#L106
Preventing navigating away during a batch update is something we agreed with @samhotchkiss.
Before going with this one, I tried a more strict path:
```
`/stats/activity/${ this.props.siteSlug }`
```
and it doesn't make a difference. There is a caveat with all these approaches (and I believe the place where I took this from—which also unnecessarily adds window handlers—also suffers from it) and is that repeated clicks in other tabs or menu items eventually allows the user to navigate without confirmation. It's something I'd like to talk to you about today since I'm not sure if it's a shortcoming of page.js or what.

Reply:
having the components know about `page` is a bit problematic. Components shouldn't know about `page`, neither about route changes. It makes everything from testing to reusing the component really difficult. And breaks the pattern of calypso's controllers managing `page`.

But rather than having a function passed down from the controller all the way down to this component, I think that the solution is not doing batch updates on the client. It adds a point of failure that is completely out of our control.

So, I agree with @dmsnell. If we don't have the API support to make batch updates on the server and notify the user of progress (and protect the process from failures) it's better not to have the feature at all. Nagging the user is never a good idea.

Reply:
The "Update all" really adds value to users since they can click once to update all plugins.

Reply:
This is not a critic on the value of the feature. But having this batch operation controlled by the client makes it extremely fragile, with a high risk of failures we can't control.

Reply:
Yup, sorry, my comment was directed to Dennis' comment above. I would prefer not to do this queue and the sequential update, but there are some limitations in the WP core updater that have been solved in the past with the sequential update. More convo at p1525452144000006-activity-log-slack

Reply:
It would be nice if we could use `lib/protect-form` here. It does exactly the same thing and the implementation is almost identical, too. The only downside is that the confirm message is hardcoded ("You have unsaved changes") and is not customizable.

If only we could extend `protectForm` with a customizable message...